src/consensus/aft/raft.h

```diff
@@ -3111,9 +3111,11 @@ namespace aft
       leader_id.reset();
 
       CCF_ASSERT_FMT(
-        !retirement_idx.has_value(),
-        "retirement_idx already set to {}",
-        retirement_idx.value());
+        !retirement_idx.has_value() ||
+          retirement_idx.value() <= state->commit_idx,
+        "unexpected attempt to pull retirement_idx forward from {} to {}",
+        retirement_idx.value(),
+        state->commit_idx);
       retirement_idx = state->commit_idx;
```

Comment:
I don't understand this fix. The assertion message suggests we shouldn't be pulling the `retirement_idx` _forward_, but it looks like the assertion condition enforces that it _does_ move forward.

Reply:
The assertion makes sure that `state->commit_idx` is not smaller than the retirement index (otherwise it should have been rolled back). In my terminology "forward" means a smaller number. There's a bug though: we can't overwrite it when it already has a value.